Sources/AblyAssetTrackingInternal/AblyWrapper/DefaultAbly.swift

```diff
@@ -129,38 +132,123 @@ public class DefaultAbly: AblyCommon {
+        channel.detach { errorInfo in
+            do{
+                try completion(errorInfo)
+            }catch{}
+        }
+    }
+
+    /**
+     * Performs the [operation] and if a "connection resume" exception is thrown it waits for the [channel] to
+     * reconnect and retries the [operation], otherwise it rethrows the exception. If the [operation] fails for
+     * the second time the exception is rethrown no matter if it was the "connection resume" exception or not.
+     */
+    private func performChannelOperationWithRetry(channel:AblySDKRealtimeChannel, operation : (AblySDKRealtimeChannel) throws ->Void) throws {
+      do {
+          logHandler?.w(message: "Trying to perform an operation on a suspended channel \(channel.name), waiting for the channel to be reconnected", error: nil)
+          try waitForChannelReconnection(channel: channel)
+          try operation(channel)
+      } catch  AblyError.connectionError(let errorInfo) {
+           if errorInfo.isConnectionResumeException(){
+            
+               logHandler?.w(message: "Connection resume failed for channel \(channel), waiting for the channel to be reconnected",
+                       error: errorInfo
+               )
+
+               do {
+                   try waitForChannelReconnection(channel: channel)
+                   try operation(channel)
+               }catch AblyError.connectionError(let secondError){
+                   logHandler?.w(message: "Retrying the operation on channel \(channel.name) has failed for the second time",
+                                          error: secondError
+                                      )
+                   throw AblyError.connectionError(errorInfo: secondError)
+               }
+           }else {
+               throw AblyError.connectionError(errorInfo: errorInfo)
+           }
+      }
+    }
+
+    /**
+     * Waits for the [channel] to change to the [ChannelState.attached] state.
+     * If this doesn't happen during the next [timeoutInMs] milliseconds, then an exception is thrown.
+     */
+    private func waitForChannelReconnection(channel:AblySDKRealtimeChannel, timeoutInMs:Int = 10_000) throws{
+        guard channel.state.toConnectionState() != .online else{
+            return
+        }
+        let blockingDispatchGroup = DispatchGroup()
+        blockingDispatchGroup.enter()
+        channel.on { stateChange in
+            if (stateChange.current.toConnectionState() == .online){
+                blockingDispatchGroup.leave()
             }
-            self?.logHandler?.e(message: "\(String(describing: Self.self)): Error while leaving the channel [id: \(trackableId)] presence", error: error)
-            completion(.failure(error.toErrorInformation()))
+            
+        }
+        let timeout: DispatchTime = .now() + .seconds(timeoutInMs / 1000)
+        
+        let result = blockingDispatchGroup.wait(timeout: timeout)
```

Comment:
I haven't looked into it to see which thread this blocks, but I don't think that we'd want to block any thread in order to perform this wait. We could instead use [`DispatchQueue.asyncAfter`](https://developer.apple.com/documentation/dispatch/dispatchqueue/2300020-asyncafter) to run code once the timeout period elapses.

Reply:
Also https://github.com/ably/ably-asset-tracking-swift/commit/248ff80d6b48c0c988fa637c81c01c9ded685831